File: 1834-single-threaded-cpu/1834-single-threaded-cpu.cpp
class Solution {
public:
    static bool cmp(pair<vector<int>,int>& a,pair<vector<int>,int>& b)
    {
        if(a.first[0]!=b.first[0])
        {
            return a.first[0]<b.first[0];
        }
        return a.first[1]<b.first[1];
    }
    vector<int> getOrder(vector<vector<int>>& tasks) {
        vector<pair<vector<int>,int>> temp;
        for(int i=0;i<tasks.size();++i)
        {
            temp.push_back({tasks[i],i});
        }
        sort(temp.begin(),temp.end(),cmp);
        long long time=temp[0].first[0],i=1,n=tasks.size();
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<>> q;
        q.push({temp[0].first[1],temp[0].second});
        vector<int> ans;
        while(i<n || !q.empty())
        {
            while(i<n && temp[i].first[0]<=time)
            {
                q.push({temp[i].first[1],temp[i].second});
                i++;
            }
            if(q.empty() && i<n)
            {
                time=temp[i].first[0];
                q.push({temp[i].first[1],temp[i].second});
                i++;
            }
            auto t1=q.top();
            q.pop();
            ans.push_back(t1.second);
            time+=t1.first;
        }
        return ans;
    }
};

File: 1834-single-threaded-cpu/1834-single-threaded-cpu.java
class Solution {
    public int[] getOrder(int[][] tasks) {
           // 1. sort the tasks by starting time
        // 2. use a minHeap to store current tasks with a custom
        // 3. terate each task, if task starting time is smaller or equal to cur, we add into the pq
        // 4. check if pq is empty, then we update current time by ith task starting time
        // 5. if not empty we update the result
        // 6. clean the minHeap
        // 7. return result
        int[][] allTasks = new int[tasks.length][3]; // index, starting time and totalTime
        Queue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[2] - b[2] == 0 ? a[0] - b[0] : a[2] - b[2]);
        for (int i = 0; i < tasks.length; i++) {
            allTasks[i][0] = i;
            allTasks[i][1] = tasks[i][0];
            allTasks[i][2] = tasks[i][1];
        }
        Arrays.sort(allTasks, (a, b) -> a[1] - b[1]);
        
        int i = 0, index = 0, curTime = 0;
        int[] result = new int[tasks.length];
        
        while (i < tasks.length) {
            // check if i th tasks starting time is smaller or equal to currenttime
            while (i < tasks.length && curTime >= allTasks[i][1]) {
                minHeap.add(allTasks[i++]);
            }
            if (minHeap.isEmpty()) {
                curTime = allTasks[i][1];
            }
            if (minHeap.isEmpty() == false) {
                int[] curTask = minHeap.remove();
                result[index++] = curTask[0];
                curTime += curTask[2];
            }
        }
        while (minHeap.isEmpty() == false) {
            int[] curTask = minHeap.remove();
            result[index++] = curTask[0];
        }
        return result;
    }
}